### PR TITLE
Propagate SecurityException from painless

### PR DESCRIPTION
**Pull request by rjernst**
Painless calls out to various Elasticsearch classes, for example when
accessing doc values. If there is an exception thrown from those
Elasticsearch classes, it is caught by painless and wrapped in a
ScriptException. However, this is not always correct, as that error
might be something that is not the users fault.

If a SecurityException is thrown while load data, for example because
an appropriate doPrivileged block was not present, we want the exception
to propagate all the way back out of the code calling painless, so that
it can be logged as a system error. This commit changes painless'
exception handling to account for SecurityException and rethrow.

relates #85217

### CODE REVIEW

**Comment by rjernst** [comment]
Note that while I was able to inject an exception for testing at runtime, I could not find a way to inject a compile time exception. I'd appreciate any suggestions there on how to test.

File: modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java [hunk 1]
```diff
@@ -425,30 +412,20 @@ protected void injectSandboxExceptions(FunctionNode irFunctionNode) {
 
             irInvokeCallNode.addArgumentNode(irLoadFieldMemberNode);
 
-            for (Class<?> throwable : new Class<?>[] {
+            irThrowNode = createCatchAndThrow(SecurityException.class, internalLocation, irTryNode);
```
Comment (stu-elastic) [inline modules/lang-painless/src/main/java/org/elasticsearch/painless/phase/PainlessUserTreeToIRTreePhase.java:415]: Can you update the comment in `injectSandboxExceptions` to document that `SecurityException` is caught and rethrown?

**Comment by stu-elastic** [comment]
`SecurityException` is whitelisted in [`java.lang.txt`](https://github.com/elastic/elasticsearch/blob/43a534e95a32ffc1ae47517b4291285acdbe4dea/modules/lang-painless/src/main/resources/org/elasticsearch/painless/java.lang.txt#L1079..L1082), with this change that exception will be logged as a system error.

What do you think about removing the `SecurityException` constructor from the whitelist?

**Comment by rjernst** [comment]
> What do you think about removing the SecurityException constructor from the whitelist?

I agree, it shouldn't be there. I'll remove.

**Comment by stu-elastic** [comment]
> I'd appreciate any suggestions there on how to test.

It'd be a lot of plumbing, but we'd need a way to pass a custom compiler for a context in [`PainlessScriptEngine`](https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/main/java/org/elasticsearch/painless/PainlessScriptEngine.java#L80), then pass in a Visitor that throws the security exception during compilation.  We already have a helper that allows [custom visitors](https://github.com/elastic/elasticsearch/blob/master/modules/lang-painless/src/main/java/org/elasticsearch/painless/Compiler.java#L265) for the tree dumper.

**Review by stu-elastic — APPROVED** [review]
Dumped the psuedo-bytecode with `Debugger.toString` and it's right:

```
   L9
    ASTORE 2
    ALOAD 2
    ATHROW
    TRYCATCHBLOCK L0 L4 L9 java/lang/SecurityException
```

**Pull request merged**